iree/task/api.c

```diff
@@ -0,0 +1,125 @@
+    "   Creates one group for each unique L2 cache group across all available\n"
+    "   cores up to the value specified by --task_topology_max_group_count.\n"
+    "   This optimizes for temporal and spatial cache locality but may suffer\n"
+    "   from oversubscription if there are other processes trying to use the\n"
+    "   same cores.\n");
+
+IREE_FLAG(
+    int32_t, task_topology_group_count, 0,
+    "Defines the total number of task system workers that will be created.\n"
+    "Workers will be distributed across cores. Specifying 0 will use a\n"
+    "heuristic defined by --task_topology_mode= to automatically select the\n"
+    "worker count and distribution.");
+
+IREE_FLAG(
+    int32_t, task_topology_max_group_count, 8,
+    "Sets a maximum value on the worker count that can be automatically\n"
+    "detected and used when --task_topology_group_count=0 and is ignored\n"
+    "otherwise.\n");
+
+// TODO(benvanik): add --task_topology_dump to dump out the current machine
+// configuration as seen by the topology utilities.
+
+//===----------------------------------------------------------------------===//
+// Task system factory functions
+//===----------------------------------------------------------------------===//
+
+IREE_API_EXPORT iree_status_t IREE_API_CALL
+iree_task_executor_create_from_flags(iree_allocator_t host_allocator,
+                                     iree_task_executor_t** out_executor) {
+  IREE_ASSERT_ARGUMENT(out_executor);
+  *out_executor = NULL;
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_task_scheduling_mode_t scheduling_mode = 0;
+  if (FLAG_task_scheduling_defer_worker_startup) {
+    scheduling_mode |= IREE_TASK_SCHEDULING_MODE_DEFER_WORKER_STARTUP;
+  }
+  if (FLAG_task_scheduling_dedicated_wait_thread) {
+    scheduling_mode |= IREE_TASK_SCHEDULING_MODE_DEDICATED_WAIT_THREAD;
+  }
+
+  iree_status_t status = iree_ok_status();
+
+  iree_task_topology_t topology;
+  iree_task_topology_initialize(&topology);
+
+  if (FLAG_task_topology_group_count != 0) {
+    iree_task_topology_initialize_from_group_count(
+        FLAG_task_topology_group_count, &topology);
+  } else if (strcmp(FLAG_task_topology_mode, "physical_cores") == 0) {
+    iree_task_topology_initialize_from_physical_cores(
+        FLAG_task_topology_max_group_count, &topology);
+  } else if (strcmp(FLAG_task_topology_mode, "unique_l2_cache_groups") == 0) {
+    iree_task_topology_initialize_from_unique_l2_cache_groups(
+        FLAG_task_topology_max_group_count, &topology);
+  } else {
+    status = iree_make_status(
+        IREE_STATUS_INVALID_ARGUMENT,
+        "one of --task_topology_group_count or --task_topology_mode must be "
+        "specified and be a valid value; have --task_topology_mode=%s.",
+        FLAG_task_topology_mode);
+  }
+
+  if (iree_status_is_ok(status)) {
+    status = iree_task_executor_create(scheduling_mode, &topology,
+                                       host_allocator, out_executor);
+  }
```

Comment:
should we return earlier if it failed?

Reply:
it needs to fall through to clean up the topology/end the trace zone regardless of whether it fails, so this lets it do that and then return the status (which may be ok or failed)